### PR TITLE
Close the client's queue when SetDSN is called with an empty DSN

### LINKED ISSUE #90: Leaking goroutines.
Merely including raven-go results in leaking goroutines. The culprit is this line:

https://github.com/getsentry/raven-go/blob/master/client.go#L307

main.go

```
package main
```

main_test.go

```
package main

import (
    "fmt"
    "os"
    "testing"

    _ "github.com/getsentry/raven-go"

    "github.com/uber/tchannel-go/testutils/goroutines"
)

func TestMain(m *testing.M) {
    exitCode := m.Run()
    if err := goroutines.IdentifyLeaks(nil); err != nil && exitCode == 0 {
        fmt.Fprintf(os.Stderr, "Found goroutine leaks on successful test run: %v", err)
        exitCode = 1
    }
    os.Exit(exitCode)
}
```

```
$ go test
PASS
Found goroutine leaks on successful test run: found unexpected goroutines:
[Goroutine 5 in state chan receive, with vendor/github.com/getsentry/raven-go.(*Client).worker on top of the stack:
goroutine 5 [chan receive]:
vendor/github.com/getsentry/raven-go.(*Client).worker(0xc8200be210)
    vendor/github.com/getsentry/raven-go/client.go:418 +0x60
created by vendor/github.com/getsentry/raven-go.newClient
    vendor/github.com/getsentry/raven-go/client.go:307 +0x193
]FAIL   test    0.913s
```

### PR DESCRIPTION
**Pull request by tylrtrmbl**
Note: This means `SetDSN(nil)` must be called after `SetDSN(dsn)`, and only exactly once.

Fixes #90

---

This change is [Reviewable](https://reviewable.io/reviews/getsentry/raven-go/93)

### CODE REVIEW

File: example/example.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ func main() {
 	if err != nil {
 		log.Fatal(err)
 	}
+	defer client.SetDSN(nil)
```
Comment (mattrobenolt) [inline example/example.go]: This won't compile since the argument is a `string`.

**Comment by mattrobenolt** [comment]
Why not just use `defer client.Close()` ?

**Comment by tylrtrmbl** [comment]
I had that same idea. However, I was inspired by this:

> If an empty DSN is passed, you should treat it as valid option which signifies disabling the client.

From the [Client API docs](https://docs.getsentry.com/hosted/clientdev/#usage-for-end-users).

It's also a little less awkward for users of the default client, who would have to do something like this:

```go
package main

import (
    sentry "github.com/getsentry/raven-go"
)

func main() {
    defer sentry.Close()
    // ...
}
```

Instead they get the more symmetric:

```go
package main

import (
    sentry "github.com/getsentry/raven-go"
)

func main() {
    sentry.SetDSN(os.Getenv("SENTRY_DSN"))
    defer sentry.SetDSN("")
    // ...
}
```

Note that I'm not actually much of a fan of the default client yet, but I was still influenced by the Client API docs.

**Comment by mattrobenolt** [comment]
I think there's a difference between initializing a client with an empty DSN, vs you already having created one, and you want to stop using it. I don't think we should overload `SetDSN` here to also indicate "stop using an existing one".

**Comment by tylrtrmbl** [comment]
I see. So you want to support:

```go
sentry.SetDSN(dsn1)
sentry.CaptureMessage("fail", nil)
sentry.CaptureMessage("all", nil)
sentry.CaptureMessage("day", nil)
sentry.CaptureMessage("and", nil)
sentry.CaptureMessage("all", nil)
sentry.CaptureMessage("night", nil)
sentry.SetDSN("")
```

and still expect all the messages to go out if the goroutine was busy.

I can support that.

**Comment by tylrtrmbl** [comment]
Would you be interested in using `Shutdown` instead of `Close`?

**Comment by mattrobenolt** [comment]
> and still expect all the messages to go out if the goroutine was busy.

Sure, I wouldn't want to stop in-flight messages.

> Would you be interested in using `Shutdown` instead of `Close`?

I don't have a strong preference. :)

**Comment by tylrtrmbl** [comment]
Oh wow. I just noticed `Close` exists already. It's merely a matter of documentation.

**Comment by mattrobenolt** [comment]
> Oh wow. I just noticed Close exists already. It's merely a matter of documentation.

lol, and I forgot.

We can/should still make the goroutine start lazy. Though, I don't think we should do it inside `SetDSN`, since it's technically possible to call `SetDSN` multiple times, in which we'll start a new goroutine for each and likely cause other issues.

We can add a `started` `bool` to the `Client` struct and just check that or something.

**Comment by tylrtrmbl** [comment]
Pushed a commit that documents `Close` in example code a little better, and puts a note to call it on all `New*` methods.

**Comment by tylrtrmbl** [comment]
You bring up an interesting point about the laziness, and I have more to say about that. I think I'll move it back to #90 though, since this pull request is now officially just a documentation boost.

**Comment by mattrobenolt** [comment]
:+1:

**Pull request closed without merging**